File: java/92.reverse-linked-list-ii.java
/*
 * @lc app=leetcode id=92 lang=java
 *
 * [92] Reverse Linked List II
 */
/**
 * Definition for singly-linked list.
 * public class ListNode {
 *     int val;
 *     ListNode next;
 *     ListNode(int x) { val = x; }
 * }
 */
class Solution {
    public ListNode reverseBetween(ListNode head, int m, int n) {
        if (head == null || head.next == null) return head;

        ListNode be4rl = head;
        int i = 1;
        if (m != 1) while (++i < m) be4rl = be4rl.next;
        else {
            be4rl = new ListNode(-1);
            be4rl.next = head;
        }

        ListNode rlp = be4rl.next;
        ListNode rlpNext = rlp.next;
        while (i++ < n) {
            ListNode tmp = rlpNext.next;
            rlpNext.next = rlp;
            rlp = rlpNext;
            rlpNext = tmp;
        }
        be4rl.next.next = rlpNext;
        be4rl.next = rlp;
        return m == 1 ? be4rl.next : head;
    }
}

File: go/92.reverse-linked-list-ii.go
/*
 * @lc app=leetcode id=92 lang=golang
 *
 * [92] Reverse Linked List II
 */

// @lc code=start
/**
 * Definition for singly-linked list.
 * type ListNode struct {
 *     Val int
 *     Next *ListNode
 * }
 */
func reverseBetween(head *ListNode, m int, n int) *ListNode {
    if head == nil || head.Next == nil {
		return head
	}
	ahead := &ListNode{0, head}
	prev, cur, next := ahead, head, head.Next
	for i := 1; i < m; i++ {
		prev, cur, next = cur, next, next.Next
	}
	prevRef, curRef := prev, cur
	for i := m; i <= n; i++ {
		cur.Next = prev
		if next != nil {
			prev, cur, next = cur, next, next.Next
		} else {
			prevRef.Next = cur
			curRef.Next = next
			return ahead.Next
		}
	}
	prevRef.Next = prev
	curRef.Next = cur
	return ahead.Next
}
// @lc code=end

